crates/sui-core/src/authority.rs

```diff
@@ -2555,17 +2560,67 @@ impl AuthorityState {
         epoch_start_timestamp_ms: CheckpointTimestamp,
     ) -> anyhow::Result<(SuiSystemState, TransactionEffects)> {
         let next_epoch = epoch_store.epoch() + 1;
-        debug!(
-            ?next_epoch,
-            computation_cost=?gas_cost_summary.computation_cost,
-            storage_cost=?gas_cost_summary.storage_cost,
-            storage_rebase=?gas_cost_summary.storage_rebate,
-            "Creating advance epoch transaction"
-        );
+
+        // Determine which protocol version to propose for the following epoch.
+        let current_protocol_version = epoch_store.committee().protocol_version;
+        let next_protocol_version = current_protocol_version + 1;
+        let next_epoch_protocol_version = if self
+            .supported_protocol_versions
+            .is_version_supported(next_protocol_version)
+        {
+            // we support the next version, check if enough other validators do as well.
+            let mut stake_aggregator: StakeAggregator<bool, true> =
+                StakeAggregator::new(Arc::new(epoch_store.committee().clone()));
+
+            let capabilities = epoch_store.get_capabilities();
+
+            let mut non_supporting = 0;
+            for cap in capabilities.iter() {
+                let authority = &cap.authority;
+                if cap
+                    .supported_protocol_versions
+                    .is_version_supported(next_protocol_version)
+                {
+                    info!(
+                        "validator {:?} supports {:?}",
+                        authority.concise(),
+                        next_protocol_version
+                    );
+                    stake_aggregator.insert_generic(*authority, true);
+                } else {
+                    info!(
+                        "validator {:?} does not support {:?}",
+                        authority, next_protocol_version
+                    );
+                    non_supporting += 1;
+                }
+            }
+
+            let total_votes = stake_aggregator.total_votes();
+            let quorum_threshold = epoch_store.committee().quorum_threshold();
+            info!(
+                ?total_votes,
+                ?quorum_threshold,
+                ?non_supporting,
+                ?next_protocol_version,
+                "support for next protocol version"
+            );
+
+            // Quorum threshold is required in order to certify the change epoch tx no matter what.
+            // We further require that we don't leave more than 3 validators behind.
+            if total_votes >= quorum_threshold && non_supporting <= 3 {
```

Comment:
There are several issues with this:
1. This is not byzantine tolerant (there can be f byzantine nodes who could all vote non_supporting
2. If anything, we should be using validator voting power instead of validator count, because validator stake can be unequally distributed
3. We are differentiating non-supporting validators and validators who didn't vote. This doesn't look very intuitive to me. Under what conditions a validator won't vote?

I think based on our discussions, that at least during the initial period, we will notify all validators about the upgrade well in advance and only pull the trigger when we are sure most will do. So to me I believe we only need the first condition, and doesn't need the second one.

Reply:
I don't believe doing an upgrade with only 2f+1 support is ever a good idea. It can lead to starting the new epoch with f dead stake.

The distinction between dissenters and abstainers is removed in #8514 - this was just an oversight on my part.

I agree that significant social coordination will be required initially, in order for upgrades to happen smoothly. This code is intended to be a safety check on top of that, not a replacement for it.